Add -r option for OpenBSD's pkg_add

moritzbuhl opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Hi, I would like to have a way to enable the -r flag in the openbsd_pkg module.

I needed this to replace the libxml on an older machine with a new patch-level version.
So there were no incompatibilities between the two port versions but due to the dependencies according to the package database I could not just replace the package in another way without first removing all the ports that required the old libxml version.

I added another yes/no switch because it is a flag to pkg_add and it can be combined with the other options on the command line.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
Replace option for openbsd_pkg

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
According to http://man.openbsd.org/pkg_add -r is used the following way:
```
     o   Replace existing packages with explicit other versions, using option
         -r.  The pkg-name ... specified on the command line are new packages
         that should replace already installed packages, with other versions
         or flavors.
...
     -r       Replace existing packages.  pkg_add will try to take every
              precaution to make sure the replacement can proceed before
              removing the old package and adding the new one, and it should
              also handle shared libraries correctly.  Among other things,
              pkg_add will refuse to replace packages as soon as it needs to
              run scripts that might fail (use -D update to force the
              replacement); pkg_add will also refuse to replace packages when
              the dependencies don't quite match (use -D updatedepends to
              force the replacement).
```

The task:
```
- name: install specific packages from update mirror
  openbsd_pkg:
    name: "{{ custom_packages }}"
    state: present
    replace: yes
  tags:
    - XXX
```

output of running a task with `replace: yes`:
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
TASK [XXX : install specific packages from update mirror] *************************************************************************
changed: [XXX] => {"changed": true, "cmd": ["pkg_add", "-r",  "XXX" ], "delta": "0:00:03.466001", "end": "2020-03-04 13:22:10.120918", "rc": 0, "start": "2020-03-04 13:22:06.654917", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
```

Task before the change:
```
- name: xxx
  shell: pkg_add -r {{ custom_packages | join(' ') }}
  environment:
    - TRUSTED_PKG_PATH: "{{ update_mirror }}"
```

moritzbuhl committed:
Add replace option

moritzbuhl committed:
adjust version

moritzbuhl committed:
fixup online editor fail

eest commented:
Hello @moritzbuhl, thanks for wanting to improve the pkg_add module :).

I have never had to use `-r` myself so I am not sure of all the implications of that flag at this point, but here are some thoughts from an initial review:

* You add `replace: yes` to the `Specify a package branch (requires at least OpenBSD 6.0)` example, I find this a bit misleading, since it might make someone think that specifying a branch (using the `%`-syntax) expects you to also add `replace: yes`, I think this should probably be it's own example to make that clear.
* In the example syntax you show in this ticket above you add `replace: yes` to a task with `state: present`, yet the code change modifies the `package_latest()` function, which is invoked if calling the module with `state: latest` (which automatically adds `-u`). What is the expected usage?
* The man page for pkg_add at http://man.openbsd.org/pkg_add seems to treat `-r`as its own specific mode of operation (the same way `install` and `upgrade` are different modes). This makes me wonder if we should treat it as such in this module also, creating a new `package_replace` function or something.

Can you supply a reproducible example of the replace operation we can use to test this work?

moritzbuhl commented:
Hi @eest, thank you for your review.

>I have never had to use -r myself so I am not sure of all the implications of that flag at this point, but here are some thoughts from an initial review:

I am not too sure about `-r` myself either. I just know that just an update did not work in my case due to all the dependencies. So I checked src for what `-r` actually does
https://github.com/openbsd/src/blob/d7e7c465daca0f7ce4af740b2043f3ecfeba01f4/usr.sbin/pkg_add/OpenBSD/PkgAdd.pm#L129
Basically says `-r` (like `-u` and `-U` ) enables `allow_replacing`. I could only find `newupdates` (from `-u` or `-U`) in https://github.com/openbsd/src/blob/1939526936671b525c38804639634545410a49b8/usr.sbin/pkg_add/OpenBSD/Dependencies.pm#L286
Whereas `-r` sets `hard_replace`, which is only used in https://github.com/openbsd/src/blob/24464ca2978098a1ff4baf17e3e553cf3389feb4/usr.sbin/pkg_add/OpenBSD/Update.pm#L164
(according to my quick search).

>You add replace: yes to the Specify a package branch (requires at least OpenBSD 6.0) example, I find this a bit misleading, since it might make someone think that specifying a branch (using the %-syntax) expects you to also add replace: yes, I think this should probably be it's own example to make that clear.

I agree and will do so.

>In the example syntax you show in this ticket above you add replace: yes to a task with state: present, yet the code change modifies the package_latest() function, which is invoked if calling the module with state: latest (which automatically adds -u). What is the expected usage?

I am not able to run my patch on a development branch of Ansible and therefore the example is completely manufactured.  I did not pay attention when trying out my change and my `remove: yes` attribute was just ignored.

>The man page for pkg_add at http://man.openbsd.org/pkg_add seems to treat -ras its own specific mode of operation (the same way install and upgrade are different modes). This makes me wonder if we should treat it as such in this module also, creating a new package_replace function or something.

As the only use of `-r` is in `Update.pm` I am not sure if it really should have it's own mode. At the same time a `pkg_add -r ansible` will just install Ansible. I am not that familiar with the Ansible way and philosophy (I honestly don't get it actually) on how to map the program options to the yaml syntax. But just adding a yes/no/[value] option for each pkg_add option does not really help either. Maybe the case is too special to be it's own option too.

>Can you supply a reproducible example of the replace operation we can use to test this work?

I am working on it:
```yml
- name: set installurl to openbsd 6.4 base
  copy:
    content: "{{ obsd_mirror_base }}\n"
    dest: /etc/installurl
  tags:
    - example_setup

- name: install the old package
  openbsd_pkg:
    name: libxml
  tags:
    - example_setup

- name: set installurl to openbsd 6.4 custom_patches
  copy:
    content: "{{ my_mirror }}\n"
    dest: /etc/installurl
  tags:
    - example_new

# TODO: Add pointing /etc/signify/openbsd-64-pkg.pub to the right key

- name: install custom built packages the old way
  shell: |
    pkg_add -r libxml
  environment:
    - TRUSTED_PKG_PATH: "{{ my_mirror }}"
  tags:
    - example_old

- name: install custom built packages the old way
  openbsd_pkg:
    name: libxml
    state: latest
    replace: yes
  tags:
    - example_new
```
So first I run `ansible-playbook playbook.yml -l myhost  -vv -t example_setup` and some vault stuff.
```
PLAYBOOK: playbook.yml ***************************************************************************************************************************
X plays in playbook.yml

TASK [Gathering facts] ***************************************************************************************************************************
task path: /data/git/mbuhl/ansible/playbook.yml:10
ok: [myhost]
META: ran handlers

TASK [role : set installurl to openbsd 6.4 base] *******************************************************************************************
task path: /data/git/mbuhl/ansible/roles/role/tasks/main.yml:27
ok: [myhost] => {"changed": false, "checksum": "05f94a6035cc96acbc3b9ed8a597f1c15e374eef", "dest": "/etc/installurl", "gid": 0, "group": "wheel", "mode": "0644", "owner": "root", "path": "/etc/installurl", "size": 27, "state": "file", "uid": 0}

TASK [role : install the old package] ******************************************************************************************************
task path: /data/git/mbuhl/ansible/roles/role/tasks/main.yml:34
changed: [myhost] => {"build": false, "changed": true, "name": ["libxml"], "state": "latest"}
META: ran handlers
META: ran handlers
```

now on myhost I get:
```
myhost mbuhl $ pkg_info libxml      
Information for inst:libxml-2.9.8p0
```

Now using the old way:
which is just the same cmd as above with the `example_old` tag.
```
TASK [Gathering facts] *********************************************************************************************************
******************
task path: /data/git/mbuhl/ansible/playbook.yml:10
ok: [myhost]
META: ran handlers

TASK [role : install custom built packages the old way] ************************************************************************
************
task path: /data/git/mbuhl/ansible/roles/role/tasks/main.yml:49
changed: [myhost] => {"changed": true, "cmd": "pkg_add -r libxml\n", "delta": "0:00:02.455686", "end": "2020-03-09 17:15:55.8231
20", "rc": 0, "start
": "2020-03-09 17:15:53.367434", "stderr": "ldconfig: /usr/local/lib/ruby/gems/2.5/gems/nokogiri-1.10.1/ports/x86_64-unknown-ope
nbsd6.4/libxml2/2.
9.9/lib: No such file or directory\nldconfig: /usr/local/lib/ruby/gems/2.5/gems/nokogiri-1.10.1/ports/x86_64-unknown-openbsd6.4/
libxslt/1.1.33/lib
: No such file or directory\nsystem(/sbin/ldconfig, -R) failed: exit(255)", "stderr_lines": ["ldconfig: /usr/local/lib/ruby/gems
/2.5/gems/nokogiri
-1.10.1/ports/x86_64-unknown-openbsd6.4/libxml2/2.9.9/lib: No such file or directory", "ldconfig: /usr/local/lib/ruby/gems/2.5/g
ems/nokogiri-1.10.
1/ports/x86_64-unknown-openbsd6.4/libxslt/1.1.33/lib: No such file or directory", "system(/sbin/ldconfig, -R) failed: exit(255)"
], "stdout": "---
-libxml-2.9.8p0 -------------------\nRemember to update /var/db/xmlcatalog", "stdout_lines": ["--- -libxml-2.9.8p0 -------------
------", "Remember
 to update /var/db/xmlcatalog"]}
META: ran handlers
META: ran handlers
```

now, checking on myhost:
```
myhost mbuhl $ pkg_info libxml 
Information for inst:libxml-2.9.10pg1
```
which is correct, as that version resembles my custom package.

Now redo the first setup (skipping)... I ensured that libxml-2.9.8p0 is installed.

And then:
```
PLAYBOOK: playbook.yml *********************************************************************************************************
******************
X plays in playbook.yml

...

TASK [Gathering facts] *********************************************************************************************************
******************
task path: /data/git/mbuhl/ansible/playbook.yml:10
ok: [myhost]
META: ran handlers

TASK [role : set installurl to openbsd 6.4 custom_patches] *********************************************************************
************
task path: /data/git/mbuhl/ansible/roles/role/tasks/main.yml:41
ok: [myhost] => {"changed": false, "checksum": "787d25016612840890831be07f47b25ab9ac6858", "dest": "/etc/installurl", "gid": 0, 
"group": "wheel", "mode": "0644", "owner": "root", "path": "/etc/installurl", "size": 28, "state": "file", "uid": 0}

TASK [role : install custom built packages the new way] ************************************************************************
************
task path: /data/git/mbuhl/ansible/roles/role/tasks/main.yml:57
ok: [myhost] => {"build": false, "changed": false, "name": ["libxml"], "state": "latest"}
META: ran handlers
META: ran handlers
```
which fails "silently" because the `-u` flag changes the behavior of the `-r` flag.
`pkg_add -umr libxml` will not update the package, `pkg_add -mr libxml` will.


Furthermore it is currently necessary to sign the packages. I can do so with a custom key (and that probably should happen if you ever plan to update machines over the Internet) or set `-D unsigned` in pkg_add or set the environment variable `TRUSTED_PKG_PATH`.

---

### Conscluion

I think it is possible to create a new mode that mostly behaves like `latest` with a `-r` option instead of `-u`. I will work on it next week.

eest commented:
> I am not too sure about `-r` myself either. I just know that just an update did not work in my case due to all the dependencies. So I checked src for what `-r` actually does
> https://github.com/openbsd/src/blob/d7e7c465daca0f7ce4af740b2043f3ecfeba01f4/usr.sbin/pkg_add/OpenBSD/PkgAdd.pm#L129
> Basically says `-r` (like `-u` and `-U` ) enables `allow_replacing`. I could only find `newupdates` (from `-u` or `-U`) in https://github.com/openbsd/src/blob/1939526936671b525c38804639634545410a49b8/usr.sbin/pkg_add/OpenBSD/Dependencies.pm#L286
> Whereas `-r` sets `hard_replace`, which is only used in https://github.com/openbsd/src/blob/24464ca2978098a1ff4baf17e3e553cf3389feb4/usr.sbin/pkg_add/OpenBSD/Update.pm#L164
> (according to my quick search).
>

That is some interesting detective work. I added a print statement to the part of Update.pm where hard_replace is checked, and could see that the code is executed no matter what flags are passed to pkg_add, so it seems Update.pm is called at all times, not just with "-u".
 
> As the only use of `-r` is in `Update.pm` I am not sure if it really should have it's own mode. At the same time a `pkg_add -r ansible` will just install Ansible. I am not that familiar with the Ansible way and philosophy (I honestly don't get it actually) on how to map the program options to the yaml syntax. But just adding a yes/no/[value] option for each pkg_add option does not really help either. Maybe the case is too special to be it's own option too.
>

What is an option and not boils down to expected usage I guess. Maby it makes sense to add a "replace: yes" option for this, the trickier part becomes how it makes the module behave regarding the execution of the pkg_add tool.

From what I can tell in your case running `pkg_add -umr` actually does not work? As in the package is not updated so you have to call `pkg_add -mr`? What is unclear then is what the expectation of "latest" is. If having "latest" bundled with "replace: yes" means removing -u and adding -r this probably means a package will _not_ be updated either, which will not be obvious to anyone looking at the playbook.

It might even make sense to make the "replace" option modify both the "present" and "latest" cases: in the case of "present" just add the "-r" flag to the invocation, and in the case of "latest" first run the "-r" operation and after that continue with the "-u" operation or something...

> ### Conscluion
> I think it is possible to create a new mode that mostly behaves like `latest` with a `-r` option instead of `-u`. I will work on it next week.

It appears there have been some remodeling of the Ansible repo, as the openbsd_pkg.py file has been removed. Looks like it moved to here: https://github.com/ansible-collections/community.general/blob/master/plugins/modules/packaging/os/openbsd_pkg.py
